### ISSUE TITLE
Idea: Compose Routers with RequestReader's

### ISSUE THREAD

**Issue by vkostyukov** [finagle/finch#232]
I remember we discussed something like this previously.

The idea is to introduce combinator `?` (like query string separator) on `Router` type that allows to compose routers and request readers together:

```scala
def getUser(userId: Int, id: Int): Future[User]
val e1: Router[Int / Int / String] = Get / "users" / int / int ? param("a")
val e2: Router[Future[User]] = Get ? param("a").as[Int] ~ param("b").as[Int] /> getUser
```

**Comment by vkostyukov** [comment]
I believe this idea fits well into the concept of heterogeneous routers (see #221). Combining these two concept will allow us to write some crazy code like this:

```scala
val api: Router[Int] = 
  (Get ? param("a").as[Int]) | // answers with given param a in the body
  (Get / "users" / int /> User.getById) |
  (Get / "echo" / string) // echo service
  
```

**Comment by rpless** [comment]
:+1: I'd like to see this. Its been in the back of my mind for a while because Finagle has extractors for params in addition to route elements and HTTP methods.

**Comment by travisbrown** [comment]
I'm planning to put some time into a demo of a Shapeless-supported heterogeneous router this weekend (in preparation for a talk at a Twitter HQ meetup in a couple of weeks)—I'll look into this as well.

**Comment by vkostyukov** [comment]
Thanks @travisbrown!

This should be relatively easy to do once we unify the composite type for both `RequestReader` and `Router`. Here is what I have in mind right now:

```scala
val a: Router[Int :: String :: HNil] = Get / int ? param("a")
val b: Router[String :: String :: Boolean :: HNil] = Get / string ? param("b") & param("c").as[Boolean]
```

Note that `&` is basically an alias for `?`. This gives us a DSL that looks pretty close to just URL.

Another crazy idea that is bothering me is to merge `RequestReader` and `Router` into a single abstraction called `Endpoint`. It should be quite possible once we close this ticket.

**Comment by ImLiar** [comment]
damn man, thats what I was looking for! :+1:

**Comment by vkostyukov** [comment]
It should be possible to do it now since we have `Smash`.

![Hulk](http://media.giphy.com/media/ErDZ73YFYLSPC/giphy.gif)

**Comment by vkostyukov** [comment]
Let me work on that.

**Comment by vkostyukov** [comment]
Just merged in #319.

It turned out, we need to make another change before start working on this one. We should probably unify `Router`s with `RequestReader`s and make them return `Future[A]` instead of just `A`. Once it's done, we should be able to compose them very easily.

This change seems pretty local to me. It doesn't touch the user-facing API so we should be fine here.

**Comment by vkostyukov** [comment]
Fixed by #360.
